### ISSUE TITLE
Mutually suggested mods are self-cancelling when both are enabled

### ISSUE THREAD

**Issue by kilbith** [minetest/minetest#3330]
Example : 

- Mod `laurel` has `hardy?` in its `depends.txt`
- Mod `hardy` has `laurel?` in its `depends.txt`

Both get enabled. Result and canceling when entering in game : 
```
ERROR[Main]: mod "laurel" has unsatisfied dependencies:  "hardy"
ERROR[Main]: mod "hardy" has unsatisfied dependencies:  "laurel"
```

**Comment by PilzAdam** [comment]
There is no way for the engine to properly resolve this, since soft depends guarantee that the mod is you depend on is loaded before yourself. Maybe the error message could be tuned, though.

**Comment by celeron55** [comment]
Instead of allowing that to happen we probably should just modify the mod spec to say that if two mods soft-depend on each other, the load order is random. The current behavior is not useful to anyone.

**Comment by celeron55** [comment]
http://irc.minetest.ru/minetest-dev/2015-11-10#i_4447455

Based on rubenwardy's comments I think it is plausible that this will stay the way it is.

**Comment by est31** [comment]
Agreed, I think it needs a better error msg, but generally behaviour is ok.

**Comment by thePalindrome** [comment]
My 2 cents here:

I've seen this sort of thing get resolved with multi-stage loading, which could be designed as such:

init.lua is called as normal

Expose minetest.register_secondary_load() which will take the mod and put it in a list of mods that require extra processing

Once all mods are loaded go through the "secondary load list" calling "process.lua" or what have you. This will allow mods that could soft depend on each other to account for that.

**Comment by celeron55** [comment]
This is already solvable in the way rubenwardy described. Just split your mod into two parts; one that provides library functionality and one that uses the functionality to add content to the game.

Complicating the modding interface with new stuff like you suggest makes no sense when it's already easily solvable by the mechanisms that we have, and is already being succesfully solved in this way in practice.

**Comment by Wuzzy2** [comment]
This is just a cyclic dependency. It's impossible to “fix” in Minetest; Minetest behaves correctly IMO.

The modders who do this are at fault here. They must redesign the mod dependencies in a way that does not create a cyclic dependency.

I think this issue could be closed, unless you maybe want to keep it open for better error messages.

**Comment by est31** [comment]
Yes, error messages are the reason why I didn't close this.

**Comment by agaran** [comment]
Well, currently both mods won't load at all throwing error messages that they hard-depend on each other, while in fact they both should load (in either order).
